uitools/cpp/Esri/ArcGISRuntime/Toolkit/ScalelineController.cpp

```diff
@@ -0,0 +1,327 @@
+/*******************************************************************************
+ *  Copyright 2012-2021 Esri
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ ******************************************************************************/
+#include "ScalelineController.h"
+
+// C++ headers
+#include <cmath>
+
+// ArcGISRuntime headers
+#include <GeometryEngine.h>
+#include <PolylineBuilder.h>
+
+namespace Esri {
+namespace ArcGISRuntime {
+namespace Toolkit {
+
+  namespace {
+    /*
+      List of "good" multipliers in which it is appropriate to scale a scaleline
+      by.
+     */
+    constexpr double MULTIPLIERS[] = {
+        1,
+        1.2,
+        1.25,
+        1.5,
+        1.75,
+        2.0,
+        2.4,
+        2.5,
+        3,
+        3.75,
+        4,
+        5,
+        6,
+        7.5,
+        8.0,
+        9.0,
+        10,
+    };
+
+    /*
+       Returns the closest multiplier to the ratio of
+       \a distance divided by \a magnitude.
+       Multipliers are from the defined set of hard-coded \c{MULTIPLIERS}.
+       If no multiplier is suitable then the first element in \c{MULTIPLIERS is returned.
+     */
+    double selectMultiplierData(double distance, double magnitude)
+    {
+      double residual = distance / magnitude;
+      return std::accumulate(
+          std::cbegin(MULTIPLIERS),
+          std::cend(MULTIPLIERS),
+          MULTIPLIERS[0],
+          [residual](double i, double m)
+          {
+            return m < residual && m > i ? m : i;
+          });
+    }
+
+    /*
+      Calculates the "magnitude" used when calculating the length of a scalebar or the number of segments. This is the
+      largest power of 10 that's less than or equal to a given \a distance.
+    */
+    double calculateMagnitude(double distance)
+    {
+      return pow(10, floor(log10(distance)));
+    }
+
+    /*
+      Returns the appropriate \l LinearUnit given a \a distance and \a unitSystem.
+      This method chooses whether miles or feet are appropriate display units for a given length in the imperial system,
+      or if kilometers of meters are appropriate display units for a given length in the metric system.
+     */
+    LinearUnit selectLinearUnit(double distance, UnitSystem unitSystem)
```

Comment:
What are the units of `distance`? For example if you pass 500 and `UnitSystem::Imperial`, does that mean 500 feet?

Reply:
Yes! I will update the docs.